test/test_pytree.py

```diff
@@ -26,6 +28,98 @@ def __init__(self, x, y):
 
 
 class TestGenericPytree(TestCase):
+    def test_aligned_public_apis(self):
+        public_apis = py_pytree.__all__
+
+        self.assertEqual(public_apis, cxx_pytree.__all__)
+
+        for name in public_apis:
+            cxx_api = getattr(cxx_pytree, name)
+            py_api = getattr(py_pytree, name)
+
+            self.assertEqual(inspect.isclass(cxx_api), inspect.isclass(py_api))
+            self.assertEqual(inspect.isfunction(cxx_api), inspect.isfunction(py_api))
+            if inspect.isfunction(cxx_api):
+                cxx_signature = inspect.signature(cxx_api)
+                py_signature = inspect.signature(py_api)
+
+                # The C++ pytree APIs provide more features than the Python APIs.
+                # The Python APIs are a subset of the C++ APIs.
+                # Check the signature of the Python API is a subset of the C++ API.
+                cxx_param_names = list(cxx_signature.parameters)
+                py_param_names = list(py_signature.parameters)
+                self.assertTrue(
+                    set(cxx_param_names).issuperset(py_param_names),
+                    msg=(
+                        f"C++ parameter(s) ({cxx_param_names}) "
+                        f"not in Python parameter(s) ({py_param_names})"
+                    ),
+                )
```

Comment:
What are the examples? Ideally we don't diverge at all in the APIs we provide, otherwise there's no swapping between python pytree and C++ pytree

Reply:
The C++ pytree provides extra `None` type handling and namespaced registry support. There are three extra features in the C++ pytree function signature:

1. Extra `namespace` (defaults to `"torch"`) keyword argument in all tree operation functions and `register_pytree_node`.
2. Extra `none_is_leaf` (defaults to `True`) keyword argument in all tree operation functions.
3. C++ pytree supports multi-tree-map, while Python pytree only supports mapping over a single pytree.

- pytorch/pytorch#91323

```python
# Extra argument `namespace`
def cxx_register_pytree_node(cls, flatten_fn, unflatten_fn, ...other args..., namespace="torch"): ...
def python_register_pytree_node(cls, flatten_fn, unflatten_fn, ...other args...): ...

# Extra argument `none_is_leaf` and `namespace`
def cxx_tree_flatten(tree, *, none_is_leaf=True, namespace="torch"): ...
def python_tree_flatten(tree): ...

# Extra argument `none_is_leaf` and `namespace` and multi-tree-map support
def cxx_tree_map(func, tree, *rests, none_is_leaf=True, namespace="torch"): ...
def python_tree_map(func, tree): ...
```

For 1 and 2, I can remove `none_is_leaf` and `namespace` arguments and hardcode them to their default values.

Reply:
Thank you